couchdb/server/compiler.py

```diff
@@ -172,3 +181,119 @@ def import_b64egg(b64str, egg_cache=None):
+        >>>
+        >>>     def is_author():
+        >>>         return self.doc['author'] == self.userctx['name']
+        >>>
+        >>>     def is_admin():
+        >>>         return '_admin' in self.userctx['roles']
+        >>>
+        >>>     def unchanged(field):
+        >>>         assert (self.olddoc is not None
+        >>>                 and self.olddoc[field] == self.newdoc[field])
+        >>>
+        >>> exports['init'] = Validate
+
+    Example of usage:
+        >>> def validate_doc_update(newdoc, olddoc, userctx):
+        >>>     init_v = require('lib/validate')['init']
+        >>>     v = init_v(newdoc, olddoc, userctx)
+        >>>
+        >>>     if v.is_admin():
+        >>>         return True
+        >>>
+        >>>     v.unchanged('author')
+        >>>     v.unchanged('created_at')
+        >>>     return True
+
+    .. versionadded:: 0.11.0
+    .. versionchanged:: 1.1.0 Available for map functions.
+    """
+    context = context or DEFAULT_CONTEXT.copy()
+    _visited_ids = []
+
+    def require(path, module=None):
+        log.debug('Looking for export objects at %s', path)
+        module = module and module.get('parent') or {}
+        new_module = resolve_module(path.split('/'), module, ddoc)
+
+        if new_module['id'] in _visited_ids:
+            log.error('Circular require calls have created deadlock!'
+                      ' DDoc id `%s` ; call stack: %r',
+                      ddoc.get('_id'), _visited_ids)
+            del _visited_ids[:]
+            raise RuntimeError('Require function calls have created deadlock')
+        _visited_ids.append(new_module['id'])
+        source = new_module['current']
+
+        module_context = context.copy()
+        module_context.update({
+            'module': new_module,
+            'exports': new_module['exports'],
+        })
+        module_context['require'] = lambda path: require(path, new_module)
+        enable_eggs = options.get('enable_eggs', False)
+        egg_cache = options.get('egg_cache', None)
+
+        try:
+            exports = maybe_export_egg(source, enable_eggs, egg_cache)
+            if exports is not None:
+                cache_to_ddoc(ddoc, new_module['id'].split('/'), exports)
+                return exports
+
+            exports = maybe_export_cached_egg(source)
+            if exports is not None:
+                return exports
+
+            bytecode = maybe_compile_function(source)
+            if bytecode is not None:
+                cache_to_ddoc(ddoc, new_module['id'].split('/'), bytecode)
+                source = bytecode
+            try:
+                exports = maybe_export_bytecode(source, module_context)
+                if exports is not None:
+                    return exports
+            except Exception as err:
+                log.exception('Failed to compile source code:\n%s',
+                              new_module['current'])
+                raise Error('compilation_error', str(err))
```

Comment:
About the error message, it will be really ugly if it is generated recursively.
e.g.
```shell
└─[iblis@imfsa Oops]% ./__main__.py
.....................................F............................................................................................
.................................................................
======================================================================
FAIL: test_fail_on_resolving_deadlock (couchdb.tests.server.compiler.DDocModulesTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/home/iblis/git/couchdb-python/couchdb/tests/server/compiler.py", line 211, in test_fail_on_resolving_deadlock
    self.assertEqual(err.args[1], 'compilation_error')
AssertionError: '(\'invalid_require_path\', \'Object None has no parent.\\n    id: None\\n    names: [\\\'helper\\\']\\n    parent
: None\\n    current: None\\n    root: {\\\'lib\\\': {\\\'utils\\\': <code object <module> at 0x805b64a08, file "<string>", line 1
>, \\\'stuff\\\': "exports[\\\'utils\\\'] = require(\\\'lib/utils\\\') \\\\nexports[\\\'body\\\'] = \\\'doc forever!\\\'", \\\'hel
per\\\': "exports[\\\'title\\\'] = \\\'best ever\\\' \\\\nexports[\\\'body\\\'] = require(\\\'lib/stuff\\\')"}}\')' != 'compilatio
n_error'
```

Hope this path make it more friendly.
```diff
diff --git a/couchdb/server/__init__.py b/couchdb/server/__init__.py
index 40a4328..260ae0d 100644
--- a/couchdb/server/__init__.py
+++ b/couchdb/server/__init__.py
@@ -149,9 +149,10 @@ class BaseQueryServer(object):
         log.exception('FatalError `%s` occurred: %s', *exc_value.args)
         if self.version < (0, 11, 0):
             id, reason = exc_value.args
-            retval = {'error': id, 'reason': reason}
+            retval = {'error': id, 'reason': str(reason)}
         else:
-            retval = ['error'] + list(exc_value.args)
+            err_name, reason = exc_value.args
+            retval = ['error', err_name, str(reason)]
         self.respond(retval)
         log.critical('That was a critical error, exiting')
         raise
@@ -166,9 +167,10 @@ class BaseQueryServer(object):
         log.exception('Error `%s` occurred: %s', *exc_value.args)
         if self.version < (0, 11, 0):
             id, reason = exc_value.args
-            retval = {'error': id, 'reason': reason}
+            retval = {'error': id, 'reason': str(reason)}
         else:
-            retval = ['error'] + list(exc_value.args)
+            err_name, reason = exc_value.args
+            retval = ['error', err_name, str(reason)]
         self.respond(retval)

     def handle_forbidden_error(self, exc_type, exc_value, xc_traceback):
diff --git a/couchdb/server/compiler.py b/couchdb/server/compiler.py
index c91a4ef..94e1d8c 100644
--- a/couchdb/server/compiler.py
+++ b/couchdb/server/compiler.py
@@ -312,7 +312,7 @@ def require(ddoc, context=None, **options):
             except Exception as err:
                 log.exception('Failed to compile source code:\n%s',
                               new_module['current'])
-                raise Error('compilation_error', str(err))
+                raise Error('compilation_error', err.args)

             raise Error('invalid_required_object', repr(new_module['current']))
         finally:
```
And some testing stuff:
```diff
diff --git a/couchdb/tests/server/compiler.py b/couchdb/tests/server/compiler.py
index fbaa2c4..fe510a0 100644
--- a/couchdb/tests/server/compiler.py
+++ b/couchdb/tests/server/compiler.py
@@ -185,6 +185,7 @@ class DDocModulesTestCase(unittest.TestCase):
         require = compiler.require(ddoc, enable_eggs=True)
         exports = require('lib/utils.py')
         self.assertEqual(exports['foo'](), 1)
+        self.assertEqual(ddoc['lib']['egg'], DUMMY_EGG)

     def test_fail_on_resolving_deadlock(self):
         ddoc = {
@@ -204,7 +205,13 @@ class DDocModulesTestCase(unittest.TestCase):
             }
         }
         require = compiler.require(ddoc)
-        self.assertRaises(exceptions.Error, require, 'lib/utils')
+        try:
+            require('lib/utils')
+        except exceptions.Error as err:
+            self.assertEqual(err.args[0], 'compilation_error')
+            self.assertEqual(err.args[1][0], 'compilation_error')
+            self.assertEqual(err.args[1][1][0], 'compilation_error')


 class EggModulesTestCase(unittest.TestCase):
diff --git a/couchdb/tests/server/qs.py b/couchdb/tests/server/qs.py
index 9937fe4..55d867f 100644
--- a/couchdb/tests/server/qs.py
+++ b/couchdb/tests/server/qs.py
@@ -55,6 +55,9 @@ class BaseQueryServerTestCase(unittest.TestCase):
         def maybe_fatal_error(func):
             def wrapper(exc_type, exc_value, exc_traceback):
                 assert exc_type is exceptions.FatalError
+                assert len(exc_value.args) == 2
+                assert isinstance(exc_value.args[0], str)
+                assert isinstance(exc_value.args[1], str)
                 return func(exc_type, exc_value, exc_traceback)
             return wrapper

@@ -99,6 +102,9 @@ class BaseQueryServerTestCase(unittest.TestCase):
         def maybe_qs_error(func):
             def wrapper(exc_type, exc_value, exc_traceback):
                 assert exc_type is exceptions.Error
+                assert len(exc_value.args) == 2
+                assert isinstance(exc_value.args[0], str)
+                assert isinstance(exc_value.args[1], str)
                 func.__self__.mock_last_error = exc_type
                 return func(exc_type, exc_value, exc_traceback)

```

The error output look like this:
```shell
└─[iblis@imfsa Oops]% ./__main__.py
.....................................E............................................................................................................................................................
======================================================================
ERROR: test_fail_on_resolving_deadlock (couchdb.tests.server.compiler.DDocModulesTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/home/iblis/git/couchdb-python/couchdb/tests/server/compiler.py", line 208, in test_fail_on_resolving_deadlock
    require('lib/utils')
  File "/usr/home/iblis/git/couchdb-python/couchdb/server/compiler.py", line 315, in require
    raise Error('compilation_error', err.args)
Error: ('compilation_error', ('invalid_require_path', 'Object None has no parent.\n    id: None\n    names: [\'helper\']\n    
parent: None\n    current: None\n    root: {\'lib\': {\'utils\': <code object <module> at 0x8066edc30, file "<string>", line 
1>, \'stuff\': "exports[\'utils\'] = require(\'./utils\') \\nexports[\'body\'] = \'doc forever!\'", \'helper\': "exports[\'title\'] = \'best 
ever\' \\nexports[\'body\'] = require(\'./stuff\')"}}'))
```